### ISSUE TITLE
How to save quantized model in resnet18_qat.py?

### ISSUE THREAD

**Issue by yangyazhou97** [Xilinx/Vitis-AI#1048]
In https://github.com/Xilinx/Vitis-AI/blob/master/src/Vitis-AI-Quantizer/vai_q_pytorch/example/resnet18_qat.py, I want to save quantized model, but it can't be loaded correctly, I notice there is "def save_checkpoint". I can load deployable.pth, but the model's parameters are not int type, I believe it should be int type after quantization.

**Comment by niuxjxlnx** [comment]
@yangyazhou97 :

In deployable.pth, model parameters are still float, but this float in different from that of model before quantization. Here the float parameters are (int*quantization scale).

**Comment by yangyazhou97** [comment]
Sorry, I don't get your point, what do you mean by int*quantization scale? I want to know the model parameter after quantization.

**Comment by yuwang-xlnx** [comment]
Hi @yangyazhou97,

The quantization process is done in forward pass and the parameters after quantization are not stored in model.state_dict().

**Comment by yangyazhou97** [comment]
Hi, @yuwang-xlnx ,
If I want to see the quantized parameters, how should I do?

**Comment by yuwang-xlnx** [comment]
> Hi, @yuwang-xlnx , If I want to see the quantized parameters, how should I do?

There is no particularly straightforward way to do this. You can refer to the code: 
https://github.com/Xilinx/Vitis-AI/blob/master/src/Vitis-AI-Quantizer/vai_q_pytorch/pytorch_binding/pytorch_nndct/nn/quantization/modules/conv_fused.py#L307

**Comment by yangyazhou97** [comment]
I added the code "print("w_quantized:{}".format(w_quantized))" in conv_fused.py, then I python resnet18_qat.py, I don't see the quantized output.

**Comment by yuwang-xlnx** [comment]
> I added the code "print("w_quantized:{}".format(w_quantized))" in conv_fused.py, then I python resnet18_qat.py, I don't see the quantized output.

Where did you add the print statement? Is your model in training mode or evaluation mode?

**Comment by yangyazhou97** [comment]
@yuwang-xlnx  I add the print statement in here, https://github.com/Xilinx/Vitis-AI/blob/master/src/Vitis-AI-Quantizer/vai_q_pytorch/pytorch_binding/pytorch_nndct/nn/quantization/modules/conv_fused.py#L307

**Comment by yuwang-xlnx** [comment]
Hi @yangyazhou97,

The statement wil print the quantized weights when your model is in training mode.

**Comment by yangyazhou97** [comment]
Hi @yuwang-xlnx, 
Can't I output the quantized weights if I use Post Training Quantization? how about Quantization Aware Training?

**Comment by yuwang-xlnx** [comment]
> Hi @yuwang-xlnx, Can't I output the quantized weights if I use Post Training Quantization? how about Quantization Aware Training?

Yes, the statement only works in Quantization Aware Training, not in Post Training Quantization.

**Comment by yangyazhou97** [comment]
I tried to output the quantization Quantization Aware Training before, but it seems that it still didn't work, the output model was still float numbers.

**Comment by yuwang-xlnx** [comment]
> I tried to output the quantization Quantization Aware Training before, but it seems that it still didn't work, the output model was still float numbers.

The fake quantization is used in Vitis AI quantizer, that means the tensors are not rellay integer numbers and reprensented by floating numbers.

**Comment by yangyazhou97** [comment]
So can't I print the saved model parameters in integer numbers and how could I testify I have really finished the quantization?

**Comment by yuwang-xlnx** [comment]
> So can't I print the saved model parameters in integer numbers and how could I testify I have really finished the quantization?

Please refer to the user guide: https://docs.xilinx.com/r/en-US/ug1414-vitis-ai/Running-Quantization-and-Getting-the-Result

**Comment by yangyazhou97** [comment]
Thanks.
